### DISCUSSION TITLE
Login error on a remote server

### DISCUSSION THREAD

**Discussion by jonas-sasin** [soofstad/react-oauth2-pkce#72] [Q&A] [answered]
Hello,
I'm using this library and everything was running smoothly until I tried to run the application on a remote server. I'm using VPN to access the machine and run the app in docker there, using nginx.
When I click on the login button, I get the following error in the console.
![image](https://user-images.githubusercontent.com/113605333/226198769-079055ac-7579-4982-ab9a-b3aaa7d4f5e6.png)

I tried running the app locally again (using docker, nginx as reverse proxy) and had no problem (basically the same setup as the remote).
I was even able to access the backend server that is behind the reverse proxy on the remote server.
Can't figure out what is the problem on the production server though.
Any tips? Thanks

Edit:
local storage state for reference
![image](https://user-images.githubusercontent.com/113605333/226201937-3dce2e1a-c3dd-4cdb-bbdb-1b04502398e4.png)

**Comment by sebastianvitterso** [comment, marked as answer]
The error says `TypeError: Cannot read properties of undefined (reading 'digest')`
That means somewhere in the code, we have `x.digest`, where `x=undefined`. 

The only place in the code where the `.digest` property is accessed, is in `pkceUtils.ts`, where line 27 reads:
```ts
const hash: ArrayBuffer = await window.crypto.subtle.digest('SHA-256', bytes) // sha256 hash it
```

This means that `window.cryptio.subtle` is undefined in you context, which according to [this StackOverflow answer], always happens in "unsecure contexts". Basically, it won't work on sites running via http, only for sites running via http**s** ([note that `localhost` is considered a "secure origin"](https://stackoverflow.com/a/46671627/11192976)). 

Is it possible that this is the issue? If so, the solution is to serve the web site/application via HTTPS.

**Reply by jonas-sasin** [reply]
Yes, this is quite certainly the cause of the issue, I just didn't expect a message like that.
Thank you for the quick and detailed explanation.

**Reply by sebastianvitterso** [reply]
I agree that the error message is suboptimal; we had not considered this facet of the application. 
@soofstad made an issue for this: https://github.com/soofstad/react-oauth2-pkce/issues/73, and we'll have a closer look at providing better feedback in cases such as this.
